[Dynamo] Guard serialization for BUILTIN_MATCH

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152729
* #152961
* #152872
* #152865
* #152730
* #152728
* #152727
* #152725

Unsupported because it uses unsupported FUNCTION_MATCH guard.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

jbschlosser committed:
[Dynamo] Guard serialization for BUILTIN_MATCH

[ghstack-poisoned]

jansel commented on torch/_dynamo/guards.py:
```diff
@@ -1776,6 +1776,9 @@ def CLOSURE_MATCH(self, guard: Guard):
             self.FUNCTION_MATCH(guard)
 
     def BUILTIN_MATCH(self, guard: Guard):
+        # don't support this in serialization because it uses unsupported FUNCTION_MATCH
+        if self.serialization_mode == "save":
+            raise RuntimeError("BUILTIN_MATCH guard cannot be serialized.")
```
Any reason the check can't be in FUNCTION_MATCH.

Builtins should be easy to support since there is a finite list of them.

jbschlosser replied:
Sounds good, I removed this check and let it pass through to the FUNCTION_MATCH error instead.

Does it makes sense in this PR to update the BUILTIN_MATCH guard logic to support the list of builtins instead of using a FUNCTION_MATCH guard or should that be done in a follow-up?

jbschlosser committed:
Update on "[Dynamo] Guard serialization for BUILTIN_MATCH"


Unsupported because it uses unsupported FUNCTION_MATCH guard.

[ghstack-poisoned]

jbschlosser committed:
Update on "[Dynamo] Guard serialization for BUILTIN_MATCH"


Unsupported because it uses unsupported FUNCTION_MATCH guard.

[ghstack-poisoned]

jansel reviewed (approved):
This seems fine, though we should support builtins since it is easy.

jbschlosser committed:
Update on "[Dynamo] Guard serialization for BUILTIN_MATCH"


Unsupported because it uses unsupported FUNCTION_MATCH guard.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "[Dynamo] Guard serialization for BUILTIN_MATCH"


Unsupported because it uses unsupported FUNCTION_MATCH guard.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
